R/hard_group_by.r

```diff
@@ -108,10 +109,41 @@ hard_group_by.data.frame <- function(df, ..., add = FALSE, .drop = FALSE) {
 #' @importFrom purrr map
 #' @importFrom purrr map
 #' @export
-hard_group_by.disk.frame <- function(df, ..., outdir=tempfile("tmp_disk_frame_hard_group_by"), nchunks = disk.frame::nchunks(df), overwrite = TRUE) {
+hard_group_by.disk.frame <- function(
+    df, 
+    ..., 
+    outdir=tempfile("tmp_disk_frame_hard_group_by"), 
+    nchunks = disk.frame::nchunks(df), 
+    overwrite = TRUE, 
+    shardby_function="hash", 
+    sort_splits=NULL, 
+    desc_vars=NULL, 
+    sort_split_sample_size=100
+  ) {
+  
   overwrite_check(outdir, overwrite)
   
   ff = list.files(attr(df, "path"))
+  stopifnot(shardby_function %in% c("hash", "sort"))
+  
+  if (shardby_function == "sort" && is.null(sort_splits)){
+    # Sample enough per chunk to generate reasonable splits
+    sample_size_per_chunk = ceiling(nchunks / disk.frame::nchunks(df)) * sort_split_sample_size
+    
+    # Sample and sort
+    sort_splits_sample <- map(df, sample_n, size=sample_size_per_chunk, replace=TRUE) %>% 
+      select(...) %>%
+      collect() %>%
+      arrange(!!!syms(...))
+    
+    # If 100 chunks, this return get 99 splits based on percentiles.
+    ntiles <- round((1:(nchunks-1)) * (nrow(sort_splits_sample) / (nchunks)))
+    
+    # Get splits. May lead to less than nchunks if duplicates are selected.
+    sort_splits <- sort_splits_sample %>% 
+      slice(ntiles) %>%
+      distinct()
```

Comment:
should it be `distinct` and then `slice`? Say we wanted 100 chunks, but only 88 distinct cuts. Then we should give a warning at that point.

Reply:
I thought it should be slice then distinct. Let's say you have 100 records in sort_splits_sample, you want to have four chunks at the end.

We want slices to be asymptotically representative of the 25th, 50th and 75th percentile ranks (by sort order) in the original dataset. 

So the logic here is to get record numbers 25, 50, 75 as the cuts. 

If we have one very common combination, then the 50th and 75th percentile could be the same, in which case we end up with one very big chunk, and only three chunks. That's where I'm thinking we need to put in a warning. And update nchunks to be 3 instead of 4 - need to find where I do that.

On the other hand, if we distinct first, then:
1. the sample weights are no longer representative of the weights they represent in the big dataset. E.g. we could have one group that has a lot of records in the big dataset, but they only count as one in the sample
2. we could end up with 88 records left, so row 75 will be 75/88 rather than 75/100

Reply:
Just one last note! I will fix it. Thanks for this awesome PR!

You just need to add `dplyr::sample_n` and `dplyr::slice` in place of `sample_n` and `slice`. Notes won't pass CRAN either.

```r
N  checking R code for possible problems (7.2s)
   hard_group_by.disk.frame: no visible binding for global variable
     'sample_n'
   hard_group_by.disk.frame: no visible global function definition for
     'slice'
   Undefined global functions or variables:
     sample_n slice
```